src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/IImportCompletionCacheService.cs

```diff
@@ -0,0 +1,17 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+#nullable enable
+
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis.Host;
+
+namespace Microsoft.CodeAnalysis.Completion.Providers.ImportCompletion
+{
+    internal interface IImportCompletionCacheService<TProject, TPortableExecutable> : IWorkspaceService
+    {
+        // PE references are keyed on assembly path.
+        IDictionary<string, TPortableExecutable> PEItemsCache { get; }
+
+        IDictionary<ProjectId, TProject> ProjectItemsCache { get; }
```

Comment:
what is TProject here.  why generic?
Why is TPortableExecutable generic?

Reply:
Those are the data we cache for each Projects/PEReferences. This cache service is used by both type and ext. method completion provider, each has their own implementation for difference types of cache'd values.

Reply:
it'd be clearer if I name them `TProjectCacheData` and `TPortableExecutableCacheData`

Reply:
can these be immutable dictionaries?